clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp

```diff
@@ -657,30 +657,40 @@ bool DependencyScanningWorker::computeDependencies(
 
   Optional<std::vector<std::string>> ModifiedCommandLine;
   llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> ModifiedFS;
-  if (ModuleName) {
-    ModifiedCommandLine = CommandLine;
-    ModifiedCommandLine->emplace_back(*ModuleName);
 
+  // If we're scanning based on a module name alone, we don't expect the client
+  // to provide us with an input file. However, the driver really wants to have
+  // one. Let's just make it up to make the driver happy.
+  if (ModuleName) {
     auto OverlayFS =
         llvm::makeIntrusiveRefCnt<llvm::vfs::OverlayFileSystem>(BaseFS);
     auto InMemoryFS =
         llvm::makeIntrusiveRefCnt<llvm::vfs::InMemoryFileSystem>();
     InMemoryFS->setCurrentWorkingDirectory(WorkingDirectory);
-    InMemoryFS->addFile(*ModuleName, 0, llvm::MemoryBuffer::getMemBuffer(""));
     OverlayFS->pushOverlay(InMemoryFS);
     ModifiedFS = OverlayFS;
+
+    SmallString<128> FakeInputPath;
+    // TODO: We should retry the creation if the path already exists.
+    llvm::sys::fs::createUniquePath(*ModuleName + "-%%%%%%%%.input",
```

Comment:
Hmm, this is bad for cached builds if the unique path ends up in the key.  Currently we don't do cached builds that use this module dependencies code path (only the TU one), but I expect we'll want to as part of supporting caching for Swift.

I'm also concerned about this (from docs):
> If \a MakeAbsolute is false the current directory will be used instead.

The current directory for the process seems like a bad choice, you probably wanted `WorkingDirectory` here.  That's still bad though, because you probably don't want to assume you can create files there.

Also, "TODO: We should retry the creation" doesn't make sense: createUniquePath itself retries in a loop until it finds a path that doesn't exist.  If that fails we don't want to try again, it probably means something is broken.

I realize this isn't very satisfying, but instead of the name `*ModuleName`, which could reasonably collide how about just choose some name unlikely to cause issues, like "__clang_scan_deps_module_input.input"; or you could append the module name to that if you really need it to be per-module (though not sure why that would matter).

Reply:
> Hmm, this is bad for cached builds if the unique path ends up in the key. Currently we don't do cached builds that use this module dependencies code path (only the TU one), but I expect we'll want to as part of supporting caching for Swift.

This path won't end up in the key. It's there to make the driver happy and to act as the "main file" for the scanner. But in this mode (by-module-name scanning), the scanner throws away the "TU" entirely and only returns the modular graph, which has no way of referencing the main file (barring bugs, e.g. not stripping the `-main-file` argument).

> I'm also concerned about this (from docs):
> 
> > If \a MakeAbsolute is false the current directory will be used instead.
> 
> The current directory for the process seems like a bad choice, you probably wanted `WorkingDirectory` here. That's still bad though, because you probably don't want to assume you can create files there.

The docs are misleading IMO. In the implementation, the only thing that happens when `MakeAbsolute == false` is that call to `sys::path::system_temp_directory()` is avoided. There's no "get CWD" call in the function. Also note that we're not creating the file in the real FS. We only put it in the temporary in-memory FS. The only objective of this piece of code is to get unique path that will not hide something important in the underlying FS.

> Also, "TODO: We should retry the creation" doesn't make sense: createUniquePath itself retries in a loop until it finds a path that doesn't exist. If that fails we don't want to try again, it probably means something is broken.

Nope, I think you're confusing this with `createUniqueFile`. The used `createUniquePath` doesn't touch the FS at all. That means it _could_ create file name that's already present in the underlying FS, but the odds are small.

> I realize this isn't very satisfying, but instead of the name `*ModuleName`, which could reasonably collide how about just choose some name unlikely to cause issues, like "__clang_scan_deps_module_input.input"; or you could append the module name to that if you really need it to be per-module (though not sure why that would matter).

Yes, using a weird file name that's unlikely to be on disk and required by the scanned module is possible. But that solution seems less robust than calling `createUniquePath()` and resolving the TODO.

Reply:
> Nope, I think you're confusing this with createUniqueFile.

You're right, that's exactly what happened 😄 

Okay, you convinced me.  Ship it.